runtime/src/main/java/org/corfudb/runtime/view/stream/ReplexStreamView.java

```diff
@@ -0,0 +1,245 @@
+                                oe instanceof ReplexOverwriteException);
+            }
+        }
+    }
+
+    /** Update the known maximum stream address, given the context.
+     *
+     * @param context   The context to use.
+     */
+    private void updateKnownMax(final ReplexStreamContext context) {
+        context.knownStreamMax = runtime.getSequencerView()
+                .nextToken(Collections.singleton(context.id), 0)
+                .getStreamAddresses().get(context.id);
+    }
+
+    /** {@inheritDoc}
+     *
+     * In Replex, we use stream addresses. We can't determine
+     * the global address until we perform the read.
+     *
+     * */
+    @Override
+    protected LogData getNextEntry(final ReplexStreamContext context,
+                                   long maxGlobal) {
+        // If we are at (or greater than, which shouldn't occur...) the known
+        // stream maximum, we need to check if theres a new stream maximum.
+        if (context.streamPointer >= context.knownStreamMax) {
+            updateKnownMax(context);
+            // There are no new entries, so we return null
+            if (context.streamPointer >= context.knownStreamMax)
+            {
+                return null;
+            }
+        }
+
+        // Read until we have exceeded the known stream maximum
+        while (context.streamPointer < context.knownStreamMax) {
+            // The address we are reading from
+            final long thisRead = context.streamPointer + 1;
+
+            // Perform the read using replex.
+            LogData ld = runtime.getAddressSpaceView()
+                    .read(context.id, thisRead, 1L)
+                    .get(context.streamPointer);
+
+            // Do we have data? If not, retry the given number of times before
+            // attempting a hole fill.
+            for (int i = 0; i < NUM_RETRIES; i++) {
+                ld = runtime.getAddressSpaceView()
+                        .read(context.id, thisRead, 1L)
+                        .get(thisRead);
+                if (ld.getType() != DataType.EMPTY) {
+                    break;
+                }
+            }
+
+            // If after we retry the data is still empty, let's hole fill.
+            while (ld.getType() == DataType.EMPTY) {
+                try {
+                    runtime.getAddressSpaceView()
+                            .fillStreamHole(context.id, thisRead);
+                } catch (OverwriteException oe) {
+                    // If we're overwritten while hole filling, that's okay
+                    // since we're going to re-read anyway
+                }
+                ld = runtime.getAddressSpaceView()
+                        .read(ID, thisRead, 1L)
+                        .get(thisRead);
+            }
+
+            // Check if we are within maxGlobal.
+            if (ld.getType() == DataType.DATA) {
+                if (ld.getGlobalAddress() > maxGlobal) {
```

Comment:
maxGlobal is the maximum global address to read to. Since replex addresses are stream-based, then we actually have to check if what we read doesn't exceed the maximum global address we requested.